pkg/controllers/nodeoverlay/controller.go

```diff
@@ -0,0 +1,256 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/apis/v1alpha1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+// Controller for validating NodeOverlay configuration and surfacing conflicts to the user
+type Controller struct {
+	kubeClient        client.Client
+	cloudProvider     cloudprovider.CloudProvider
+	instanceTypeStore *InstanceTypeStore
+
+	temporaryStore *InstanceTypeStore
+}
+
+func (c *Controller) Name() string {
+	return "nodeoverlay.controller"
+}
+
+// NewController constructs a controller for node overlay validation
+func NewController(kubeClient client.Client, cp cloudprovider.CloudProvider, instanceTypeStore *InstanceTypeStore) *Controller {
+	return &Controller{
+		kubeClient:        kubeClient,
+		cloudProvider:     cp,
+		instanceTypeStore: instanceTypeStore,
+		temporaryStore:    NewInstanceTypeStore(),
+	}
+}
+
+// Reconcile validates that all node overlays don't have conflicting requirements
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	overlayList := &v1alpha1.NodeOverlayList{}
+	nodePoolList := &v1.NodePoolList{}
+	if err := c.kubeClient.List(ctx, overlayList); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeoverlays, %w", err)
+	}
+	err := c.kubeClient.List(ctx, nodePoolList)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodepools, %w", err)
+	}
+
+	overlayWithRuntimeValidationFailure := map[string]error{}
+	overlaysWithConflict := []string{}
+	c.temporaryStore = NewInstanceTypeStore()
+	nodePoolToInstanceTypes := map[string][]*cloudprovider.InstanceType{}
+
+	for _, np := range nodePoolList.Items {
+		its, err := c.cloudProvider.GetInstanceTypes(ctx, &np)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+		}
+		nodePoolToInstanceTypes[np.Name] = its
+	}
+
+	overlayList.OrderByWeight()
+	for i := range overlayList.Items {
+		if err := overlayList.Items[i].RuntimeValidate(ctx); err != nil {
+			overlayWithRuntimeValidationFailure[overlayList.Items[i].Name] = err
+			continue
+		}
+
+		// Due to reserved capacity type offering being dynamically injected as part of the GetInstanceTypes call
+		// We will need to make sure we are validating against each nodepool to make sure. This will ensure that
+		// overlays that are targeting reserved instance offerings will be able to apply the offering.
+		for _, np := range nodePoolList.Items {
```

Comment:
nit: would it make more sense to just iterate directly over the keys and values of `nodePoolsToInstanceTypes` here?